### PR TITLE
 Adds handlers to internalApi

### PR DESCRIPTION
**Pull request by sachithKay**

### CODE REVIEW

File: components/inbound-endpoints/org.wso2.carbon.inbound.endpoint/src/main/java/org/wso2/carbon/inbound/endpoint/internal/http/api/InternalAPIHandler.java [hunk 1]
```diff
@@ -0,0 +1,27 @@
+package org.wso2.carbon.inbound.endpoint.internal.http.api;
```
Comment (malakaganga) [inline components/inbound-endpoints/org.wso2.carbon.inbound.endpoint/src/main/java/org/wso2/carbon/inbound/endpoint/internal/http/api/InternalAPIHandler.java]: Shall we add the license?

Comment (sachithKay) [reply components/inbound-endpoints/org.wso2.carbon.inbound.endpoint/src/main/java/org/wso2/carbon/inbound/endpoint/internal/http/api/InternalAPIHandler.java]: done

File: components/inbound-endpoints/org.wso2.carbon.inbound.endpoint/src/main/java/org/wso2/carbon/inbound/endpoint/internal/http/api/InternalAPIHandler.java [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.wso2.carbon.inbound.endpoint.internal.http.api;
+
+import org.apache.synapse.MessageContext;
+
+public interface InternalAPIHandler {
```
Comment (milindaperera) [inline components/inbound-endpoints/org.wso2.carbon.inbound.endpoint/src/main/java/org/wso2/carbon/inbound/endpoint/internal/http/api/InternalAPIHandler.java]: Let's add a description

File: components/inbound-endpoints/org.wso2.carbon.inbound.endpoint/src/main/java/org/wso2/carbon/inbound/endpoint/internal/http/api/ConfigurationLoader.java [hunk 3]
```diff
@@ -161,6 +160,49 @@ public static void loadInternalApis(String apiFilePath) {
         }
     }
 
+    private static void setHandlers(OMElement apiElement, InternalAPI api) {
+
+        List<InternalAPIHandler> handlerList = new ArrayList<>();
+        OMElement handlersOM = apiElement.getFirstElement();
```
Comment (milindaperera) [inline components/inbound-endpoints/org.wso2.carbon.inbound.endpoint/src/main/java/org/wso2/carbon/inbound/endpoint/internal/http/api/ConfigurationLoader.java]: Get element by QName

File: components/inbound-endpoints/org.wso2.carbon.inbound.endpoint/src/main/java/org/wso2/carbon/inbound/endpoint/internal/http/api/InternalAPIHandler.java (hunk 2 above)
Comment (sachithKay) [reply components/inbound-endpoints/org.wso2.carbon.inbound.endpoint/src/main/java/org/wso2/carbon/inbound/endpoint/internal/http/api/InternalAPIHandler.java]: done

File: components/inbound-endpoints/org.wso2.carbon.inbound.endpoint/src/main/java/org/wso2/carbon/inbound/endpoint/internal/http/api/ConfigurationLoader.java (hunk 3 above)
Comment (sachithKay) [reply components/inbound-endpoints/org.wso2.carbon.inbound.endpoint/src/main/java/org/wso2/carbon/inbound/endpoint/internal/http/api/ConfigurationLoader.java]: done

**Comment by Nirothipan** [comment]
@sachithKay There was a unit test which loads the internal API xml and do some testing . Shall we add a sample handler for it and check whether it is getting loaded properly.

File: components/inbound-endpoints/org.wso2.carbon.inbound.endpoint/src/main/java/org/wso2/carbon/inbound/endpoint/internal/http/api/ConfigurationLoader.java [hunk 4]
```diff
@@ -161,6 +160,49 @@ public static void loadInternalApis(String apiFilePath) {
         }
     }
 
+    private static void setHandlers(OMElement apiElement, InternalAPI api) {
+
+        List<InternalAPIHandler> handlerList = new ArrayList<>();
+        OMElement handlersOM = apiElement.getFirstChildWithName(new QName("handlers"));
```
Comment (isudana) [inline components/inbound-endpoints/org.wso2.carbon.inbound.endpoint/src/main/java/org/wso2/carbon/inbound/endpoint/internal/http/api/ConfigurationLoader.java]: handlers QName can be defined at the top (same as API_Q)

Comment (isudana) [inline components/inbound-endpoints/org.wso2.carbon.inbound.endpoint/src/main/java/org/wso2/carbon/inbound/endpoint/internal/http/api/ConfigurationLoader.java]: rename handlersOM to handlersElement

File: components/inbound-endpoints/org.wso2.carbon.inbound.endpoint/src/main/java/org/wso2/carbon/inbound/endpoint/internal/http/api/ConfigurationLoader.java [hunk 5]
```diff
@@ -161,6 +160,49 @@ public static void loadInternalApis(String apiFilePath) {
...
+        if (handlersOM != null) {
+            Iterator<OMElement> handlers = handlersOM.getChildElements();
+
+            while (handlers.hasNext()) {
+                OMElement handlerOM = handlers.next();
```
Comment (isudana) [inline components/inbound-endpoints/org.wso2.carbon.inbound.endpoint/src/main/java/org/wso2/carbon/inbound/endpoint/internal/http/api/ConfigurationLoader.java]: handlerOM -> handlerElement

File: components/inbound-endpoints/org.wso2.carbon.inbound.endpoint/src/main/java/org/wso2/carbon/inbound/endpoint/internal/http/api/ConfigurationLoader.java [hunk 6]
```diff
@@ -161,6 +160,49 @@ public static void loadInternalApis(String apiFilePath) {
...
+                if (handlerOM.getAttribute(new QName("name")) != null) {
+                    String handlerName = handlerOM.getAttributeValue(new QName("name"));
```
Comment (isudana) [inline components/inbound-endpoints/org.wso2.carbon.inbound.endpoint/src/main/java/org/wso2/carbon/inbound/endpoint/internal/http/api/ConfigurationLoader.java]: user NAME_ATT which is already defined.

File: components/inbound-endpoints/org.wso2.carbon.inbound.endpoint/src/main/java/org/wso2/carbon/inbound/endpoint/internal/http/api/ConfigurationLoader.java [hunk 7]
```diff
@@ -161,6 +160,49 @@ public static void loadInternalApis(String apiFilePath) {
...
+                    if (handlerOM.getAttribute(new QName("class")) != null) {
```
Comment (isudana) [inline components/inbound-endpoints/org.wso2.carbon.inbound.endpoint/src/main/java/org/wso2/carbon/inbound/endpoint/internal/http/api/ConfigurationLoader.java]: use CLASS_Q which is already defined.

File: components/inbound-endpoints/org.wso2.carbon.inbound.endpoint/src/main/java/org/wso2/carbon/inbound/endpoint/internal/http/api/ConfigurationLoader.java [hunk 8]
```diff
@@ -161,6 +160,49 @@ public static void loadInternalApis(String apiFilePath) {
...
+                        String handlerClass = handlerOM.getAttributeValue(new QName("class"));
+                        InternalAPIHandler handler = createHandler(handlerClass);
+                        handler.setName(handlerName);
+                        handlerList.add(handler);
+                    } else {
+                        handleException("Class not defined in one or more handlers");
```
Comment (isudana) [inline components/inbound-endpoints/org.wso2.carbon.inbound.endpoint/src/main/java/org/wso2/carbon/inbound/endpoint/internal/http/api/ConfigurationLoader.java]: print a descriptive error message with the handler name

File: components/inbound-endpoints/org.wso2.carbon.inbound.endpoint/src/main/java/org/wso2/carbon/inbound/endpoint/internal/http/api/ConfigurationLoader.java [hunk 9]
```diff
@@ -161,6 +160,49 @@ public static void loadInternalApis(String apiFilePath) {
         }
     }
 
+    private static void setHandlers(OMElement apiElement, InternalAPI api) {
```
Comment (isudana) [inline components/inbound-endpoints/org.wso2.carbon.inbound.endpoint/src/main/java/org/wso2/carbon/inbound/endpoint/internal/http/api/ConfigurationLoader.java]: populateHandlers() is a better name that setHandlers()

File: components/inbound-endpoints/org.wso2.carbon.inbound.endpoint/src/main/java/org/wso2/carbon/inbound/endpoint/internal/http/api/ConfigurationLoader.java (hunk 4 above)
Comment (sachithKay) [reply components/inbound-endpoints/org.wso2.carbon.inbound.endpoint/src/main/java/org/wso2/carbon/inbound/endpoint/internal/http/api/ConfigurationLoader.java]: Fixed

Comment (sachithKay) [reply components/inbound-endpoints/org.wso2.carbon.inbound.endpoint/src/main/java/org/wso2/carbon/inbound/endpoint/internal/http/api/ConfigurationLoader.java]: Fixed

File: components/inbound-endpoints/org.wso2.carbon.inbound.endpoint/src/main/java/org/wso2/carbon/inbound/endpoint/internal/http/api/ConfigurationLoader.java (hunk 5 above)
Comment (sachithKay) [reply components/inbound-endpoints/org.wso2.carbon.inbound.endpoint/src/main/java/org/wso2/carbon/inbound/endpoint/internal/http/api/ConfigurationLoader.java]: Fixed

File: components/inbound-endpoints/org.wso2.carbon.inbound.endpoint/src/main/java/org/wso2/carbon/inbound/endpoint/internal/http/api/ConfigurationLoader.java (hunk 6 above)
Comment (sachithKay) [reply components/inbound-endpoints/org.wso2.carbon.inbound.endpoint/src/main/java/org/wso2/carbon/inbound/endpoint/internal/http/api/ConfigurationLoader.java]: Fixed

File: components/inbound-endpoints/org.wso2.carbon.inbound.endpoint/src/main/java/org/wso2/carbon/inbound/endpoint/internal/http/api/ConfigurationLoader.java (hunk 7 above)
Comment (sachithKay) [reply components/inbound-endpoints/org.wso2.carbon.inbound.endpoint/src/main/java/org/wso2/carbon/inbound/endpoint/internal/http/api/ConfigurationLoader.java]: Fixed

File: components/inbound-endpoints/org.wso2.carbon.inbound.endpoint/src/main/java/org/wso2/carbon/inbound/endpoint/internal/http/api/ConfigurationLoader.java (hunk 8 above)
Comment (sachithKay) [reply components/inbound-endpoints/org.wso2.carbon.inbound.endpoint/src/main/java/org/wso2/carbon/inbound/endpoint/internal/http/api/ConfigurationLoader.java]: Fixed

File: components/inbound-endpoints/org.wso2.carbon.inbound.endpoint/src/main/java/org/wso2/carbon/inbound/endpoint/internal/http/api/ConfigurationLoader.java (hunk 9 above)
Comment (sachithKay) [reply components/inbound-endpoints/org.wso2.carbon.inbound.endpoint/src/main/java/org/wso2/carbon/inbound/endpoint/internal/http/api/ConfigurationLoader.java]: Fixed

**Pull request merged**